Tests/Test.cs

```diff
@@ -0,0 +1,327 @@
+using ProductItem = (Guid ProductId, int Quantity);
+using PricedProductItem = (Guid ProductId, int Quantity, decimal Price);
+
+// Commands
+public abstract record ShoppingCartCommand
+{
+    public record AddProduct(
+        ProductItem ProductItem,
+        Guid? ClientId = null
+    ): ShoppingCartCommand;
+
+    public record RemoveProduct(
+        PricedProductItem ProductItem
+    ): ShoppingCartCommand;
+
+    public record Confirm(
+        Guid ClientId
+    ): ShoppingCartCommand;
+
+    public record Cancel: ShoppingCartCommand;
+
+    private ShoppingCartCommand() { }
+}
+
+public abstract record ShoppingCartEvent
+{
+    public record Opened(
+        Guid? ClientId,
+        DateTimeOffset OpenedAt
+    ): ShoppingCartEvent;
+
+    public record ProductAdded(
+        PricedProductItem ProductItem,
+        DateTimeOffset AddedAt
+    ): ShoppingCartEvent;
+
+    public record ProductRemoved(
+        PricedProductItem ProductItem,
+        DateTimeOffset RemovedAt
+    ): ShoppingCartEvent;
+
+    public record Confirmed(
+        Guid ClientId,
+        DateTimeOffset ConfirmedAt
+    ): ShoppingCartEvent;
+
+    public record Cancelled(
+        DateTimeOffset CanceledAt
+    ): ShoppingCartEvent;
+
+    private ShoppingCartEvent() { }
+}
+
+public abstract record ShoppingCart
+{
+    public record Empty: ShoppingCart;
+
+    public record Pending(ProductItems ProductItems): ShoppingCart;
+
+    public record Closed: ShoppingCart;
+
+    private ShoppingCart() { }
+}
+
+// Primary constructor & TimeProvider
+// Then Collection expressions and advanced pattern matching
+public class ShoppingClassDecider(
+    IProductPriceCalculator priceCalculator,
+    TimeProvider timeProvider
+)
+{
+    public ShoppingCartEvent[] Decide(ShoppingCartCommand command, ShoppingCart state) =>
+        (state, command) switch
+        {
+            (Empty, AddProduct (var productItem, var clientId)) =>
+            [
+                new Opened(clientId, Now),
+                new ProductAdded(priceCalculator.Calculate(productItem).Single(), Now)
+            ],
+
+            (Pending, AddProduct (var productItem, _)) =>
+            [
+                new ProductAdded(priceCalculator.Calculate(productItem).Single(), Now)
+            ],
+
+            (Pending pending, RemoveProduct((var productId, var toRemove, var price) pricedProductItem)) =>
+                pending.ProductItems.TryGetValue($"{productId}{price}", out var quantity) && quantity >= toRemove
+                    ?
+                    [
+                        new ProductRemoved(pricedProductItem, Now)
+                    ]
+                    : throw new InvalidOperationException("Not enough Products!"),
+
+            (Pending, Confirm(var clientId)) =>
+            [
+                new Confirmed(clientId, Now)
+            ],
+
+            (Pending, Cancel) =>
+            [
+                new Cancelled(Now)
+            ],
+
+            (Closed, Confirm) => [],
```

Comment:
collapse two cases to one ?

Reply:
You cannot, as C# doesn't have union types. I can do just (Closed, _) which would catch also other operations.